### ISSUE TITLE
print to file 

### ISSUE THREAD

**Issue by QianruZhou333** [seveno4/EURISKO#2]
Apologize for opening a new issue, but it is the only way I could think of to getting in touch with you. I want to thank you for your work, it is truly remarkable and highly appreciated. I am trying to study the Eurisko code and reimplement it as well, but the current code print output on display windows, which is difficult to track for me. I am trying to print the messages to file, but as I am new to lisp, I could not find a way to do it successfully (as shown in [Stack overflow](https://stackoverflow.com/questions/78645353/write-text-i-o-stream-to-file-with-interlisp-medley) ). Is that possible that you could provide some suggestions on that, any advises will be gratefully accepted. Thank you very much.

**Comment by masinter** [comment]
I added a note referencing this question to https://github.com/Interlisp/medley/issues/1405#issuecomment-2185277334 
the Medley issue for the FAQ (if not directly addressed with the medley primer.

There are some migration libraries in the Medley sources which need revival, but those might be useful if you want to reimplement Interlisp code starting with another Common Lisp implementation.

**Comment by masinter** [comment]
https://github.com/Interlisp/medley/issues/1774

I finally tried the example you gave and was surprised; dont think this behavior is right, or well-documented.

But this is not a eurisko problem and this issue should be closed.

**Comment by QianruZhou333** [comment]
@masinter Thank you very much for your reply and for your answer in stack overflow. It is true that it is not a problem of Eurisko. I have finally print it to file by reopen and  (CLOSEF ...) everytime I write to file. 

I am new to lisp. I hope it is not too bold to ask why do you think it is not a right behaviour to print it to file? It is the only way I could think of to study what exactly has Eurisko learnt after running for hundreds of tasks.

**Comment by QianruZhou333** [comment]
May I ask another question, after executing Eurisko to task 584, it reports an error "Worth is an unbound variable", why does it happen and how to debug it?

**Comment by QianruZhou333** [comment]
Another question: after executing Eurisko for, like, 500 tasks, how can I find out which units has Eurisko discovered that are interesting?

**Comment by seveno4** [comment]
Sorry for not responding sooner.

Yes, I've seen that bug before but I'm not sure why it happens or how to debug it.

How Eurisko is meant to be used is still rather unclear to me but I think that the 'Units' with the most 'Worth' are the most useful to the system and therefore the most interesting... You may find this [article](http://blog.funcall.org/docs/am-eurisko/Eurisko_Heuristics_III.pdf) interesting as it had the most public information about Eurisko before the source code was discovered.

**Comment by white-flame** [comment]
Do note that this is not a released piece of software. It's an earlier in-development snapshot that existed from before the TCS challenge and a year or two before the major papers (2 years younger than the one linked above), but it's the only copy that's ever been found. Having pored through it all in detail, there are some obvious logical & syntactic problems in there that would blow it up if executed.

This is digital archaeology more than anything else. :-) The later "full" versions were likely only inside a Lisp world image (basically like a VM RAM snapshot/suspend state), and not in standalone source files. None of those have been found, and Lenat said for decades that he didn't have a copy anymore.

Regarding its use, at least for the Traveller stuff, it was literally just Lenat looking through the overnight logs and his personal evaluation of which heuristics to include in the next night's runs. He described coming up with the winning fleet as about 60% him, and 40% Eurisko in terms of work done.

The system has its own notions of Worth in how often they apply, and its own technical notions of interestingness (AM-like things like the edge cases of a set, or if 2 sets have a very small intersection those intersecting elements might be interesting). That's bound to be divergent from what a human will find interesting in terms of domain application and novelty. Browsing through the set of Units was supposed to be RLL's job, but that never was completed nor integrated with Eurisko. As of now, it's just basically pulling up units at the REPL using the existing accessor functions as far as I can tell. You should be able to call the Snazzy* functions manually as well, giving it stuff to display in the GUI.

IMO there are 3 main repos to look at, depending on what you want.
- https://github.com/white-flame/eurisko - The original unmodifed source as discovered, which isn't that useful but the wiki has some info, and more is welcome.
- https://github.com/seveno4/EURISKO - The one you're at now. This has seveno4's modifications that let it run on Medley with GUI. This is what you want if you're interested in running it.
- https://github.com/white-flame/eurisclo - My in-progress direct port of it to Common Lisp. It's not in a working state, but is the most human-readable version. You can glean some knowledge from the source code there, which can help you figure out how to run the working seveno4's version from here.

**Comment by Angular-Angel** [comment]
> @masinter Thank you very much for your reply and for your answer in stack overflow. It is true that it is not a problem of Eurisko. I have finally print it to file by reopen and (CLOSEF ...) everytime I write to file.
> 
> I am new to lisp. I hope it is not too bold to ask why do you think it is not a right behaviour to print it to file? It is the only way I could think of to study what exactly has Eurisko learnt after running for hundreds of tasks.

Can you tell me what commands you used to make it print? I can't get it to do that at all.

**Comment by QianruZhou333** [comment]
@Angular-Angel the code looks like this:
`(PROGN (SETQ MYFILE (OPENSTREAM 'myfile 'APPEND))`
		`(TERPRI MYFILE)`
		`(PRIN1 "Modified multiplicities of elements of examples of " MYFILE)`
		`(PRIN1 CurUnit MYFILE)`
		`(PRIN1 (QUOTE s) MYFILE)`
		`(PRIN1 " and as a result added " MYFILE)`
		`(PRIN1 (LENGTH NewValues) MYFILE)`
		`(PRIN1 " new examples." MYFILE)`
		`(TERPRI MYFILE)`
               ` (CLOSEF? MYFILE))`
basically I open and close the file 'myfile' everytime I want to write to it.

**Comment by QianruZhou333** [comment]
@white-flame @seveno4  Thank you so much for your reply and everything. Sorry I did not reply any sooner. 
I did find 'RLL' file in the https://www.saildart.org/[AM,DBL], which like you said is Lenat's own log for running Eurisko, and I found the "Goldbach conjection" in it!! So it is real......

**Comment by Angular-Angel** [comment]
So, it turns out there's an easy way to just print all output to a file: 
```
(DRIBBLE 'MYFILE)
All prints will be sent to MYFILE. You can close the dribble file with (DRIBBLE)
```
Credit to @seveno4. Much easier!

**Comment by Angular-Angel** [comment]
> @white-flame @seveno4 Thank you so much for your reply and everything. Sorry I did not reply any sooner. I did find 'RLL' file in the [https://www.saildart.org/[AM,DBL]](https://www.saildart.org/%5BAM,DBL%5D), which like you said is Lenat's own log for running Eurisko, and I found the "Goldbach conjection" in it!! So it is real......

Also, which file has the RLL code in it? There are a couple files called RLL, but neither has code in it to my knowledge - one is a description of what they want from RLL, the other is a log of output from Eurisko.

**Comment by Angular-Angel** [comment]
Also, to anyone who wants to discuss Eurisko more generally, let me invite you to the discord: https://discord.gg/vhsmVCwgvK
